### PR TITLE
add struct support

### PR DESCRIPTION
**Pull request by zhfnjust**

### CODE REVIEW

File: src/scryptTypes.ts [hunk 1]
```diff
@@ -293,4 +293,24 @@ export class OpCodeType extends ScryptType {
...
+
+type MAP<T> =   {
+  [key: string]: T;
+}
+
+type StructObject = MAP<SingletonParamType>;
+
+export class Struct {
+
+  constructor(o: StructObject) {
```
Comment (xhliu) [inline src/scryptTypes.ts:306]: Add comments explaining why using proxy pattern here.

**Review by xhliu — APPROVED** [review]
🥇 Looking good.

Just 1 suggestion.

File: test/fixture/person.scrypt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    public function main(struct Person p, int x, bool isMale) {
+
+        require(p.isMale == isMale);
+        require(p.age > x);
+        require(! p.isMale);
+
+        struct Person c = this.incAge(p);
+        require(c.age == 34);
```
Comment (freedomhero) [inline test/fixture/person.scrypt:29]: struct Person c = this.incAge(p);
require(c.age == 34);
require(p.age == 33);


p = this.incAge(p);
require(p.age == 34);

Comment (zhfnjust) [reply test/fixture/person.scrypt:29]: 6896aa57a96d35ef0203f2c957914ffb2bb37688

File: src/scryptTypes.ts (hunk 1 above)
Comment (zhfnjust) [reply src/scryptTypes.ts:306]: 2352e264f2932aed216201fcca5babad88b82ab9

**Pull request merged**